File: day9/part1.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

type Point struct {
	X int
	Y int
}

func abs(x int) int {
	if x < 0 {
		return -x
	}
	return x
}

func main() {
	file, err := os.Open("input.txt")
	if err != nil {
		panic(err)
	}
	defer file.Close()

	points := make([]Point, 0)
	scanner := bufio.NewScanner(file)

	for scanner.Scan() {
		var x, y int
		fmt.Sscanf(scanner.Text(), "%d,%d", &x, &y)
		points = append(points, Point{x, y})
	}

	maxArea := 0
	n := len(points)

	for i := 0; i < n; i++ {
		for j := i + 1; j < n; j++ {
			area := (abs(points[i].X-points[j].X) + 1) * (abs(points[i].Y-points[j].Y) + 1)
			if area > maxArea {
				maxArea = area
			}
		}
	}

	fmt.Println(maxArea)
}

File: day9/part1.c
#include <stddef.h>
#include <stdio.h>
#include <stdlib.h>

typedef struct {
  int x;
  int y;
} Point;

int main(void) {
  FILE *f = fopen("input.txt", "r");
  if (!f) {
    perror("input.txt");
    return 1;
  }

  size_t cap = 1024;
  size_t n = 0;
  Point *points = malloc(cap * sizeof(Point));

  while (1) {
    Point p;
    if (fscanf(f, "%d,%d", &p.x, &p.y) != 2)
      break;

    if (n == cap) {
      cap *= 2;
      points = realloc(points, cap * sizeof(Point));
    }
    points[n++] = p;
  }
  fclose(f);

  long long max_area = 0;

  for (size_t i = 0; i < n; i++) {
    for (size_t j = i + 1; j < n; j++) {
      long long area = (long long)(abs(points[i].x - points[j].x) + 1) *
                       (long long)(abs(points[i].y - points[j].y) + 1);

      if (area > max_area)
        max_area = area;
    }
  }

  printf("%11lld\n", max_area);

  free(points);
  return 0;
}
